server/resource/caldav_service.go

```diff
@@ -0,0 +1,426 @@
+
+
+	_, _, mailerError := cs.Mailer.DoAction(Outcome{}, actionRequestParameters)
+	if mailerError  != nil{
+		log.Error("Unable To Send mail", mailerError)
+	}
+
+	log.Info("resource created ", rpath)
+	return &res, nil
+}
+
+func (cs *CalDavStorage) UpdateResource(rpath, content string) (*data.Resource, error) {
+	a, err := cs.haveAccess(rpath, "write")
+	if err != nil {
+		log.Error(err, "failed to get Access [" + rpath + "]")
+		return nil, err
+	}
+	if ! a {
+		log.Info("no access to collection [" + rpath + "]")
+		return nil, nil
+	}
+	c := base64.StdEncoding.EncodeToString([]byte(content))
+	err = cs.cruds["calendar"].UpdateResource(rpath, c)
+	if err != nil {
+		log.Error(err, "failed to update ", rpath)
+		return nil, err
+	}
+
+	res := data.NewResource(rpath, &PGResourceAdapter{db: cs.cruds, resourcePath: rpath,  UserID: cs.UserID})
+	_ = res.GetPropertyValue("VEVENT", "ATTENDEE")
+	title     := res.GetPropertyValue("VEVENT", "SUMMARY")
+	start     := res.StartTimeUTC()
+	end       := res.EndTimeUTC()
+	subject   := "Updated invitation: " + title + " @ " + start.Weekday().String() + " " + start.Month().String() + " " + strconv.Itoa(start.Day()) + ", " + strconv.Itoa(start.Year()) + " " + strconv.Itoa(start.Hour()) + " - " + strconv.Itoa(end.Hour()) + " (" + start.Location().String() + ") (" + cs.Email + ")"
+
+	actionRequestParameters := make(map[string]interface{})
+	actionRequestParameters["to"] = cs.Email
+	actionRequestParameters["subject"] = subject
+	actionRequestParameters["from"] = "daptin.no-reply"
+	actionRequestParameters["body"] = content
+
+
+	_, _, mailerError := cs.Mailer.DoAction(Outcome{}, actionRequestParameters)
+	if mailerError  != nil{
+		log.Error("Unable To Send mail", mailerError)
+	}
+
+	log.Info("resource updated ", rpath)
+	return &res, nil
+}
+
+func (cs *CalDavStorage) DeleteResource(rpath string) error {
+	a, err := cs.haveAccess(rpath, "admin")
+	if err != nil {
+		log.Error(err, "failed to get Access [" + rpath + "]")
+		return  err
+	}
+	if ! a {
+		log.Info("no access to collection [" + rpath + "]")
+		return nil
+	}
+
+	err = cs.cruds["calendar"].DeleteCalendarEvent(cs.UserID, rpath)
+	if err != nil {
+		log.Info("failed to delete resource ", rpath, " ", err.Error())
+		return err
+	}
+
+	return nil
+}
+
+func (cs *CalDavStorage) isResourcePresent(rpath string) bool {
+	rrpath, err := cs.cruds["calendar"].GetRpath(cs.UserID)
+
+	if err != nil {
+		return false
+	}
+
+	if rrpath == rpath {
+		return true
+	}
+
+	return false
+}
+
+type PGResourceAdapter struct {
+	db map[string]*DbResource
+	resourcePath string
+	UserID       int64
+}
+
+func (pa *PGResourceAdapter) CalculateEtag() string {
+	if pa.IsCollection() {
+		return ""
+	}
+
+	return fmt.Sprintf(`"%x%x"`, pa.GetContentSize(), pa.GetModTime().UnixNano())
+}
+
+func (pa *PGResourceAdapter) haveAccess(perm string) (bool, error) {
+	permInst := pa.db[USER_ACCOUNT_TABLE_NAME].GetObjectPermissionById("calendar", pa.UserID)
```

Comment:
GetObjectPermissionById takes in objectType as first param (calendar here) and the id of the object (calendar row id) as second param, so this will fail as UserId is being passed here